File: lab3/python/gridgraph.py
import sys
import re
from typing import List, Set, FrozenSet, Union, Optional

from edge import Edge
from graph import Graph
from point import Point


class GridGraph(Graph[Point]):
    """
    GridGraph is a 2D-map encoded as a bitmap, or an N x M matrix of characters.

    Some characters are passable, others denote obstacles.
    A verte is a point in the bitmap, consisting of an x- and a y-coordinate.
    This is defined by the helper class `Point`.
    You can move from each point to the eight point around it.
    The edge costs are 1.0 (for up/down/left/right) and sqrt(2) (for diagonal movement).
    The graph can be read from a simple ASCII art text file.
    """
    grid: List[str]

    # Characters from Moving AI Lab:
    #   . - passable terrain
    #   G - passable terrain
    #   @ - out of bounds
    #   O - out of bounds
    #   T - trees (unpassable)
    #   S - swamp (passable from regular terrain)
    #   W - water (traversable, but not passable from terrain)
    # Characters from http://www.delorie.com/game-room/mazes/genmaze.cgi
    #   | - +  walls
    #   space  passable
    allowedChars = ".G@OTSW +|-"
    passableChars = ".G "

    # The eight directions, as points.
    directions = [
        Point(x, y)
        for x in [-1, 0, 1] for y in [-1, 0, 1]
        if not (x == y == 0)
    ]

    def width(self) -> int:
        return len(self.grid[0])

    def height(self) -> int:
        return len(self.grid)

    def __init__(self, graph: Optional[Union[str, List[str]]] = None):
        if graph:
            self.init(graph)

    def init(self, graph: Union[str, List[str]]):
        """
        Initialises the graph with edges from a text file,
        or from a grid of characters.
        The file describes the graph as ASCII art,
        in the format of the graph files from the Moving AI Lab.
        """
        if isinstance(graph, str):
            with open(graph, encoding="utf-8") as IN:
                graph = [
                    line for line in IN.read().splitlines() 
                    if re.match("^[" + GridGraph.allowedChars + "]+$", line)
                ]
        self.grid = graph
        for row in self.grid:
            if len(row) != self.width():
                raise ValueError("Malformed grid, row widths doesn't match.");

    def passable(self, p: Point) -> bool:
        """Returns true if you're allowed to pass through the given point."""
        return (
            0 <= p.x < self.width() and
            0 <= p.y < self.height() and
            self.grid[p.y][p.x] in self.passableChars
        )

    def vertices(self) -> FrozenSet[Point]:
        # Note: this is inefficient because it calculates the set each time.
        return frozenset(
            p for y in range(self.height()) 
            for x in range(self.width())
            if self.passable(p := Point(x, y))
        )

    def outgoingEdges(self, v: Point) -> List[Edge[Point]]:
        return [
            edge
            # We consider all directions...
            for dir in self.directions
            # ...compute the edge in that direction...
            for edge in [Edge(v, v.add(dir), dir.euclideanNorm())]
            # ...and keep the ones with passable target.
            if self.passable(edge.end)
        ]

    def isWeighted(self) -> bool:
        return True

    def guessCost(self, v: Point, w: Point) -> float:
        """
        Returns the guessed best cost for getting from one point to another.
        (the Euclidean distance between the points)
        """
        #---------- TASK 4: Guessing the cost, GridGraph ---------------------#
        # TODO: Replace these lines with your solution!
        raise NotImplementedError()
        #---------- END TASK 4 -----------------------------------------------#

    def parseVertex(self, s: str) -> Point:
        """
        Parse a point from a string representation.
        For example, a valid string representation is "39:18".
        """
        return Point.parse(s)

    def drawGraph(self, 
                maxWidth: int,
                maxHeight: int,
                start: Optional[Point] = None, 
                goal: Optional[Point] = None, 
                solution: Optional[List[Edge[Point]]] = None,
            ) -> str:
        pathPoints: Set[Point] = set()
        if solution:
            for e in solution:
                pathPoints.add(e.start)
                pathPoints.add(e.end)

        lines: List[str] = []
        for y, row in enumerate(self.grid):
            if y >= maxHeight:
                lines.append("(truncated)")
                break
            line: List[str] = []
            for x, p in enumerate(row):
                if y == 0 and x >= maxWidth - 10:
                    line.append(" (truncated)")
                    break
                if x >= maxWidth:
                    break
                point = Point(x, y)
                line.append(
                    'S' if point == start else
                    'G' if point == goal else
                    '*' if point in pathPoints else p
                )
            lines.append("".join(line))
        return "\n".join(lines)

    def __str__(self) -> str:
        """
        Returns a string representation of this graph, including some random points and edges.
        """
        return (
            f"Bitmap graph of dimensions {self.width()} x {self.height()} pixels.\n" +
            self.drawGraph(100, 25) +
            "\n\nRandom example points with outgoing edges:\n" +
            self.exampleOutgoingEdges(8)
        )


if __name__ == '__main__':
    _, file = sys.argv
    graph = GridGraph(file)
    print(graph)

File: lab3/java/GridGraph.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.Stream;

/**
 * GridGraph is a 2D-map encoded as a bitmap, or an N x M matrix of characters.
 *
 * Some characters are passable, others denote obstacles.
 * A node is a point in the bitmap, consisting of an x- and a y-coordinate.
 * This is defined by the helper class `Point`.
 * You can move from each point to the eight point around it.
 * The edge costs are 1.0 (for up/down/left/right) and sqrt(2) (for diagonal movement).
 * The graph can be read from a simple ASCII art text file.
 */

public class GridGraph implements Graph<Point> {
    private char[][] grid;

    // Characters from Moving AI Lab:
    //   . - passable terrain
    //   G - passable terrain
    //   @ - out of bounds
    //   O - out of bounds
    //   T - trees (unpassable)
    //   S - swamp (passable from regular terrain)
    //   W - water (traversable, but not passable from terrain)
    // Characters from http://www.delorie.com/game-room/mazes/genmaze.cgi
    //   | - +  walls
    //   space  passable
    // Note: "-" must come last in allowedChars, because we use it unescaped in a regular expression.

    private static final String allowedChars = ".G@OTSW +|-";
    private static final String passableChars = ".G ";

    // The eight directions, as points.
    private static final Point[] directions =
        IntStream.rangeClosed(-1, 1).boxed().flatMap(x ->
            IntStream.rangeClosed(-1, 1).boxed().flatMap(y ->
                Stream.of(new Point(x, y)).filter(p -> !p.equals(Point.ORIGIN))
            )
        ).toArray(Point[]::new);

    public int width() {
        return this.grid[0].length;
    }

    public int height() {
        return this.grid.length;
    }

    GridGraph() {}
    GridGraph(String graph) throws IOException {
        this.init(graph);
    }
    GridGraph(char[][] graph) {
        this.init(graph);
    }

    /**
     * Initialises the graph with edges from a text file.
     * The file describes the graph as ASCII art,
     * in the format of the graph files from the Moving AI Lab.
     */
    @Override
    public void init(String graph) throws IOException {
        init(Files.lines(Paths.get(graph))
                .filter(line -> line.matches("^[" + allowedChars + "]+$"))
                .map(String::toCharArray)
                .toArray(char[][]::new)
        );
    }

    /**
     * Initialises the graph from a grid of characters.
     */
    public void init(char[][] graph) {
        this.grid = graph;
        for (char[] row : this.grid) {
            if (row.length != this.width())
                throw new IllegalArgumentException("Malformed grid, row widths doesn't match.");
        }
    }

    /**
     * Returns true if you're allowed to pass through the given point.
     */
    private boolean passable(Point p) {
        return (
            0 <= p.x && p.x < this.width() && 
            0 <= p.y && p.y < this.height() && 
            GridGraph.passableChars.indexOf(this.grid[p.y][p.x]) >= 0
        );
    }

    @Override
    public Set<Point> vertices() {
        // Note: this is inefficient because it calculates the set each time.
        HashSet<Point> vertices = new HashSet<>();
        for (int y = 0; y < this.height(); y++) {
            for (int x = 0; x < this.width(); x++) {
                Point p = new Point(x, y);
                if (this.passable(p))
                    vertices.add(p);
            }
        }
        return Collections.unmodifiableSet(vertices);
    }

    @Override
    public List<Edge<Point>> outgoingEdges(Point p) {
        return 
            // We consider all directions...
            Arrays.stream(GridGraph.directions)
            // ...compute the edge in that direction...
            .map(dir -> new Edge<>(p, p.add(dir), dir.euclideanNorm()))
            // ...keep the ones with passable target...
            .filter(edge -> this.passable(edge.end))
            // ...and return them as a list.
            .collect(Collectors.toList());
    }

    public boolean isWeighted() {
        return true;
    }

    /**
     * Returns the guessed best cost for getting from one point to another.
     * (the Euclidean distance between the points)
     */
    @Override
    public double guessCost(Point v, Point w) {
        //---------- TASK 4: Guessing the cost, GridGraph -----------------//
        return v.subtract(w).euclideanNorm();
        //---------- END TASK 4 -------------------------------------------//
    }

    /**
     * Parse a point from a string representation.
     * For example, a valid string representation is "39:18".
     */
    @Override
    public Point parseVertex(String s) {
        return Point.parse(s);
    }

    /**
     * Returns a graphical string representation of the grid, with the given path shown.
     */
    @Override
    public String drawGraph(int maxWidth, int maxHeight, Point start, Point goal, List<Edge<Point>> solution) {
        Set<Point> pathPoints = new HashSet<>();
        if (solution != null) {
            solution.stream().map(e -> e.start).forEach(pathPoints::add);
            solution.stream().map(e -> e.end).forEach(pathPoints::add);
        }

        ArrayList<String> lines = new ArrayList<>();
        for (int y = 0; y < this.height(); y++) {
            if (y >= maxHeight) {
                lines.add("(truncated)");
                break;
            }
            ArrayList<String> line = new ArrayList<>();
            for (int x = 0; x < this.width(); x++) {
                if (y == 0 && x >= maxWidth - 10) {
                    line.add(" (truncated)");
                    break;
                }
                if (x >= maxWidth) {
                    break;
                }
                char p = this.grid[y][x];
                Point point = new Point(x, y);
                line.add(
                    point.equals(start) ? "S" :
                    point.equals(goal) ? "G" :
                    pathPoints.contains(point) ? "*" : String.valueOf(p)
                );
            }
            lines.add(String.join("", line));
        }
        return String.join("\n", lines);
    }

    /**
     * Returns a string representation of this graph, including some random points and edges.
     */
    @Override
    public String toString() {
        return (
            String.format("Bitmap graph of dimensions %d x %d pixels.\n", this.width(), this.height()) +
            this.drawGraph(100, 25) +
            "\n\nRandom example points with outgoing edges:\n" + 
            this.exampleOutgoingEdges(8)
        );
    }

    public static void main(String[] args) throws IOException {
        if (args.length != 1) throw new IllegalArgumentException();
        GridGraph graph = new GridGraph(args[0]);
        System.out.println(graph);
    }
}

